### PR TITLE
feat: add support for transparency

### LINKED ISSUE #131: Add support for RGBA?
Hello!

Would be cool if support could be added for `rgba()`. E.g: `pastel "rgba(0, 0, 0, 0.1)"`. Could perhaps output something like this:
```
Name: black                    
Hex: #000000  
RGB: rgb(0, 0, 0)              
RGBA: rgba(0, 0, 0, 0.1)            
HSL: hsl(0, 0.0%, 0.0%)        
Opacity: 0.1
```

Backstory is that I find myself copying colors from design tools which can give my something rgba values in my clipboard. Would be nice to use directly with pastel without changing the output from the clipboard.

### PR DESCRIPTION
**Pull request by superhawk610**
This PR adds support for parsing and displaying partially transparent colors. It adds a new function `Color::composite` which implements simple [alpha compositing](https://en.wikipedia.org/wiki/Alpha_compositing), and uses this to enhance `Canvas::draw_rect` to support painting partially transparent pixels over existing pixels. This is used to enable drawing color previews for transparent colors (see example below).

Additionally, this PR enhances the parser to support parsing all common alpha formats, including:

- `RGBA`
- `RRGGBBAA`
- `rgba()`
- `hsla()`

The `lab` parser also supports specifying alpha. The alpha value may be provided as a double or percentage, and may be slash, comma, or space delimited; all of the following are valid:

- `rgba(255 127 0 0.4)`
- `rgba(255, 127, 0, 0.4)`
- `rgba(255, 127, 0, 40%)`

For backwards compatibility, `rgb()` and `hsl()` also support alpha values. You can check the `parse_alpha_syntax` test to see all supported formats.

All `Color::to_*_string` methods now include the alpha value when it's not `1.0`, and omit it when it is.

Closes #131.

## Open Questions

This PR doesn't update the simplified format used when displaying colors with `pastel format` (the alpha channel value will be displayed, but the color itself will be fully opaque). It also doesn't add support for rendering colored text with partial transparency.

## Example

```
$ pastel color 6633997f
```

image

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -547,6 +591,35 @@ impl Color {
...
+    pub fn composite(&self, source: &Color) -> Color {
+        let backdrop = self.to_rgba();
+        let source = source.to_rgba();
+
+        // Composite A over B (see https://en.wikipedia.org/wiki/Alpha_compositing)
+        //
+        //   αo = αa + αb(1 - αa)
+        //   Co = Ca * αa + Cb * αb(1 - αa)
+        //        -------------------------
+        //                   αo
```
Comment (sharkdp) [inline src/lib.rs:605]: Maybe
```suggestion
        //        Ca * αa + Cb * αb(1 - αa)
        //   Co = -------------------------
        //                   αo
```

File: src/lib.rs [hunk 2]
```diff
@@ -141,14 +141,22 @@ impl Color {
...
+            a_prefix = if self.alpha == 1.0 { "" } else { "a" },
+            h = self.hue.value(),
+            s = 100.0 * self.saturation,
+            l = 100.0 * self.lightness,
+            space = space,
+            a = if self.alpha == 1.0 {
+                "".to_string()
+            } else {
+                format!(",{space}{:.3}", self.alpha, space = space)
+            }
```
Comment (sharkdp) [inline src/lib.rs:159]: Maybe we could set
```rust
(a, a_prefix) = if self.alpha == 1.0 { … } else { … }
```
before the `format!` call to avoid having to duplicate the `self.alpha == 1.0` condition?

Concerning the `.3` digits for the alpha value: does that allow us to properly reconstruct all alpha values from 00 to ff? That is: can we properly serialize and deserialize i/255.0 for i = 0 … 255?

File: src/parser.rs [hunk 3]
```diff
@@ -93,54 +137,57 @@ fn parse_hex(input: &str) -> IResult<&str, Color> {
 }
 
 fn parse_numeric_rgb(input: &str) -> IResult<&str, Color> {
-    let (input, prefixed) = opt(tag("rgb("))(input)?;
+    let (input, prefixed) = opt(alt((tag("rgb("), tag("rgba("))))(input)?;
```
Comment (sharkdp) [inline src/parser.rs:140]: Hm. So do we allow `rgba(1,2,3)` or `rgb(1,2,3,30%)` as well?

File: src/lib.rs [hunk 4]
```diff
@@ -547,6 +591,35 @@ impl Color {
             .mix(&C::from_color(other), fraction)
             .into_color()
     }
+
+    /// Alpha composite two colors, placing the second over the first.
+    pub fn composite(&self, source: &Color) -> Color {
```
Comment (sharkdp) [inline src/lib.rs:596]: Nice!

File: src/cli/output.rs [hunk 5]
```diff
@@ -35,7 +35,7 @@ impl Output<'_> {
         let text_position_x: usize = checkerboard_size + 2 * config.padding;
         let text_position_y: usize = 0;
 
-        let mut canvas = Canvas::new(checkerboard_size, 51, config.brush);
+        let mut canvas = Canvas::new(checkerboard_size, 60, config.brush);
```
Comment (sharkdp) [inline src/cli/output.rs:38]: Why is this necessary?

**Review by sharkdp — COMMENTED** [review]
Thank you very much!

Great PR description, implementation and tests!

File: src/lib.rs (hunk 2 above)
Comment (superhawk610) [reply src/lib.rs:159]: I couldn't figure out a good solution for decimal/hex alpha values. Most hex alphas are irrational, so there's no good way to reliably convert back and forth between a decimal value. For example, hex `80` is `128 / 255`, or `0.501960784...` (non-terminating). I went with fixed precision here so I could at least write some form of test for the parser:

```rust
    // hex alpha doesn't line up nicely with decimal precision,
    // so just compare the debug output (3 digit precision)
    assert_eq!(
        format!("{:?}", Some(rgba(10, 0, 0, 0.502))),
        format!("{:?}", parse_color("0a000080"))
    );
```

but I'm not super happy with that either. I don't really have a good idea for how to allow alpha values to be converted freely between decimal and hex, or if there even is a good way to do that at all. Three digit precision seemed like a reasonable choice for human readability (personally I've only ever seen alphas out to 3 digits for very small values like `0.005`, and even that's rare, most alphas are 1-2 digit precision).

I'm open to ideas?

File: src/parser.rs (hunk 3 above)
Comment (superhawk610) [reply src/parser.rs:140]: Here's the spec according to the most recent [CSS Color Module draft](https://www.w3.org/TR/css-color-4/) from W3C:

```
rgb() = rgb( [<percentage> | none]{3} [ / [<alpha-value> | none] ]? ) |
  rgb( [<number> | none]{3} [ / [<alpha-value> | none] ]? )
<alpha-value> = <number> | <percentage>
```

> For legacy reasons, [rgb()](https://www.w3.org/TR/css-color-4/#funcdef-rgb) also supports a [legacy color syntax](https://www.w3.org/TR/css-color-4/#legacy-color-syntax) that separates all of its arguments with commas

> Also, an rgba() [legacy color syntax](https://www.w3.org/TR/css-color-4/#legacy-color-syntax) also exists, with an identical grammar and behavior to [rgb()](https://www.w3.org/TR/css-color-4/#funcdef-rgb).

The `a`-suffixed variants of `rgb` and `hsl` have been deprecated in that Level 4 draft, but to maintain compatibility they essentially function identically to their non-suffixed counterparts. While `rgba()` with 3 values and `rgb()` with 4 values look a bit odd, they're accepted by most modern browsers (checked Chrome and Firefox) so I'd lean towards allowing them.

File: src/cli/output.rs (hunk 5 above)
Comment (superhawk610) [reply src/cli/output.rs:38]: Transparent colors will use the extended `rgba()`/`hsla()` notation, which takes an additional 8 characters for alpha values with 3 digit precision:

```
# no transparency ---

  ▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀  RGB: rgb(255, 215, 0)
  ▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀  HSL: hsl(51, 100.0%, 50.0%)
--------------------------------------------------|
                                                  51 chars

# with transparency ---

  ▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀  RGB: rgba(255, 215, 0, 0.123)
  ▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀  HSL: hsla(51, 100.0%, 50.0%, 0.123)
------------------------------------------------------|
                                                      55 chars
```

I think it could be reduced from `60` to `55`?

File: src/parser.rs [hunk 6]
```diff
@@ -427,3 +471,91 @@ fn parse_named_syntax() {
...
 }
+
+#[test]
+fn parse_alpha_syntax() {
+    // hex
+    assert_eq!(Some(rgba(255, 0, 0, 1.0)), parse_color("ff0000ff"));
+    assert_eq!(Some(rgba(255, 0, 0, 1.0)), parse_color("#ff0000ff"));
+
+    // rgb/rgba
+    assert_eq!(Some(rgba(10, 0, 0, 1.0)), parse_color("rgb(10,0,0/1)"));
```
[the remark this answers is not in the document]
Comment (sharkdp) [reply src/parser.rs:482]: Hm, yes. Let's maybe do that. Personally, I think it can easily be confused with division.

File: src/cli/output.rs (hunk 5 above)
Comment (sharkdp) [reply src/cli/output.rs:38]: Ah, I see! To be honest, I don't remember where the 51 came from. I believe it was a worst-case bound (long color names?), not a rough guess. But I'm not sure anymore.

**Review by superhawk610 — COMMENTED** [review]
Alright, updated with a few changes:

- remove support for the `/ alpha` notation
- updated the CLI help for `color` to include alpha transparency notation
- omit trailing zeroes when printing decimal alpha values

I originally had a function like this to implement max precision:

```rust
fn max_precision(value: f64, max_precision: u32) -> String {
    // (...snip...)
    format!("{}", rounded)
}
```

but I ended up going with a struct `MaxPrecision` that implements `std::fmt::Display` to save the `String` allocation. I'm still fairly new to Rust so let me know your thoughts on this (if any 🙂).

File: src/lib.rs (hunk 2 above)
Comment (sharkdp) [reply src/lib.rs]: > I couldn't figure out a good solution for decimal/hex alpha values. Most hex alphas are irrational, so there's no good way to reliably convert back and forth between a decimal value. For example, hex 80 is 128 / 255, or 0.501960784... (non-terminating).

Hm, yes. But that's not necessarily a problem, right? If alpha values are restricted to 8 bit anyway, then a few digits might be enough. What I meant was: 3 digits should be fine, because this succeeds (using Python for simplicity):
```python
for alpha_int in range(256):
    alpha_float = alpha_int / 255
    alpha_string_3digits = f"{alpha_float:.3f}"
    parsed_float = float(alpha_string_3digits)
    parsed_int = round(parsed_float * 255)

    assert alpha_int == parsed_int
```
… while two digits would not be enough.

Maybe we could write a unit test that does something similar?

Comment (superhawk610) [reply src/lib.rs]: Good point, I guess I was getting too tied up with being able to do

```rust
assert_eq!(Some(rgba(10, 0, 0, 0.502)), parse_color("0a000080"));
```

instead of comparing `Debug` outputs. I think you're right that this isn't a real problem, so long as we can reliably go back and forth between hex 0..255 and three decimal precision. I'll translate your suggestion as a unit test 👍

File: src/lib.rs [hunk 7]
```diff
@@ -1669,4 +1775,19 @@ mod tests {
...
+        // alpha values. This test insures that is sufficient to interchange
+        // back and forth between hex (values 0..256) and float (values 0..=1),
+        // e.g. hex `80` is float `0.502`, which parses to hex `80`, and so on.
+        for alpha_int in 0..256 {
+            let alpha_float: f64 = alpha_int as f64 / 255.0;
+            let alpha_string_3digits = format!("{}", MaxPrecision::wrap(3, alpha_float));
+            let parsed_float = alpha_string_3digits.parse::<f64>().unwrap();
+            let parsed_int = (parsed_float * 255.0).round() as i32;
+            assert_eq!(alpha_int, parsed_int);
+        }
```
Comment (sharkdp) [inline src/lib.rs:1791]: Well.. this is good. But it would be even better if we could implement this test using `Color` functionality only. I.e. create 256 different `Color`s with appropriate alpha values, serialize them to a string, parse them again, and see if the result is the same as the input.

File: src/lib.rs [hunk 8]
```diff
@@ -1669,4 +1783,19 @@ mod tests {
...
+        // alpha values. This test insures that is sufficient to "roundtrip"
+        // from hex (0 < n < 255) to float (0 < n < 1) and back again,
+        // e.g. hex `80` is float `0.502`, which parses to hex `80`, and so on.
+        for alpha_int in 0..255 {
+            let hex_string = format!("#000000{:02x}", alpha_int);
+            let parsed_from_hex = hex_string.parse::<Color>().unwrap();
+            let rgba_string = parsed_from_hex.to_rgb_float_string(Format::Spaces);
+            let parsed_from_rgba = rgba_string.parse::<Color>().unwrap();
+            assert_eq!(hex_string, parsed_from_rgba.to_rgb_hex_string(true));
+        }
```
Comment (superhawk610) [inline src/lib.rs:1799]: Alright, I think this should provide proper coverage. This asserts that:

- for every hex alpha value, 0 < n < 255 (`00` - `fe`)
- parsing a color with this alpha, in hex notation
- outputting that color as `rgba()`, with alpha as decimal
- parsing _that_ color
- outputting that color using hex notation
- generates the same value as the original hex input string

For example, `#00000080` parses to a color that stringifies to `rgba(0, 0, 0, 0.502)`, that in turn parses to a color that stringifies to `#00000080`.

I also threw in an `impl FromStr for Color` that calls `parser::parse_color`, that's not required, just thought it may be useful.

Comment (sharkdp) [reply src/lib.rs:1799]: Awesome, thank you!

**Review by sharkdp — APPROVED** [review]
Thank you!
